Use spdlog for logging

Montellese opened the pull request:
## Description
This is a new try for #11833. One of the main reasons why I abandoned #11833 was because `spdlog` didn't support some of the things we needed like fixed width formatting for thread id and log level. In the meantime this has been added to `spdlog`.

This replaces our homemade logging implementation with [spdlog](https://github.com/gabime/spdlog) which is a header-only C++11 logging library which depends on the recently introduced [{fmt}](https://github.com/fmtlib/fmt) library (so we can use the exact same formatting style).

I only tested this on Windows and Linux so far. It should compile on Android but I didn't runtime test it. OSX/iOS is completely untested. Please take a look at the following platform-specific interface implementations:
* POXIS: https://github.com/xbmc/xbmc/commit/0d5c4f72fe96dc4c554b69fc75d5b3406802de55#diff-51de34646b2699bf489acf948be92df8
* Darwin: https://github.com/xbmc/xbmc/commit/0d5c4f72fe96dc4c554b69fc75d5b3406802de55#diff-fbfc294a64371ab26ee25ab6e1606385
* Android: https://github.com/xbmc/xbmc/commit/0d5c4f72fe96dc4c554b69fc75d5b3406802de55#diff-4de9b444ed50f7831abc986d45ef8fda
* Windows: https://github.com/xbmc/xbmc/commit/0d5c4f72fe96dc4c554b69fc75d5b3406802de55#diff-00603a9e2c8c8d5ab5b5bcb7acd5ca34

For Ubuntu there is a `libspdlog-dev` package available for Xenial at http://packages.ubuntu.com/xenial/libdevel/libspdlog-dev but I wasn't able to figure out which version it actually is because it says 1.6 but that's not a valid release of spdlog. Unfortunately it is not at all compatible with version `1.5.0` I used for this PR because it is missing several include files and some essential methods in the logger implementation.

## Motivation and Context
The advantage of a standard logging implementation is that we don't have to maintain our own implementation and they are well tested.
spdlog comes with a set of sinks that cover all of our use cases (console/file/adb logging/VS/... logging) and more that we might want to use in the future (file rotation, ...). Furthermore it supports named loggers which should make logging in the code much easier and also make the log file easier to read because every class/instance follows the same guidelines (right now every class logs their class name differently or not at all). As an example I've implemented a named logger for the `CWebServer` class in 1082652398f5f0b77deda5b8e2904690df0bf396.

Using templates I also managed to get rid of the difference between `Log` and `LogW` which makes the interface of `CLog` easier to use.

## How Has This Been Tested?
Local testing on windows.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

## TODOs
- [x] make sure the integration into `depends` is correct
- [x] make sure the integration into CMake is correct
- [x] find a solution for `LOGINFO` vs. `LOGNOTICE`
- [x] find a solution for `LOGSEVERE` vs. `LOGERROR` / `LOGFATAL`
- [x] decide whether to `#include <spdlog/spdlog.h>` in `utils/log.h` or not
- [x] decide whether to enable or disable internal spdlog (header only) build by default

Montellese committed:
[cmake/depends/win32] add spdlog

Montellese committed:
[network] CWebServer: make use of new logging functionality

Montellese committed:
remove unnecessary trailing newline from log messages

Montellese committed:
[utils] CLog: replace custom logging implementation with spdlog

Montellese commented:
As can be seen from the TODOs in the PR description there are a few open questions:
* please take a look at my `depends` integration. Since I mainly work on Windows I'm not really comfortable with the `depends` buildsystem and probably did something (or more) wrong.
* please take a look at my `CMake` integration especially the `FindSpdlog.cmake` and the `ENABLE_INTERNAL_SPDLOG` option
* what do we want to do about the fact that `spdlog` doesn't offer a `notice` log level?
* initially I had `utils/log.h` include `spdlog/spdlog.h` directly but then I realized that this introduces a dependency from `python_binding` on `spdlog` in `CMake` and I wasn't sure how to fix that. I had to manually add the dependency to https://github.com/xbmc/xbmc/pull/17498/commits/c7f253792e5706a2f9285a5c0119806bfcc84e0f#diff-ef9013ae2ec9c1cd85464b3c714d0871 to solve this. Is there a better way to do this? What I did with e00d257bb59edca136288bc50b98761ee4391ea5 is to forward declare everything we absolutely need from `spdlog` in `utils/log.h` and move everything else to `utils/log.cpp`. But this also has a drawback: every class with uses the new `CLog::Get()` to get a specific logger (like `CWebServer` in 6949ab66bc0d2fa28a8a9b74ab577545559d21a5 now also has to manually `#include <spdlog/spdlog.h>` (see https://github.com/xbmc/xbmc/pull/17498/commits/e00d257bb59edca136288bc50b98761ee4391ea5) which I don't really like either.

lrusak commented:
jenkins build this please

Montellese committed:
[settings/lib] make use of new logging functionality

ksooo reviewed (approved):
PVR change lgtm.

Montellese commented:
Concerning the last question I didn't manage to get all platforms working properly with the forward declaring version of `utils/log.h`. Therefore I will revert back to the initial version where `utils/log.h` will contain all necessary includes of `spdlog`. Nonetheless I've introduced a `utils/logtypes.h` which contains some forward declarations to not pollute the headers of classes using `spdlog` loggers directly.

Now the question is how to properly specify the dependendy of `python_binding` on `spdlog`. The same already applies to `python_binding` requiring `fmt` but I guess this always works due to other dependencies.

lrusak commented on xbmc/utils/log.cpp:
```diff
@@ -9,208 +9,198 @@
+Logger CLog::m_defaultLogger = nullptr;
 
-CLog::CLog() = default;
+#if defined(TARGET_ANDROID)
```
instead of all this ifdeffing could you not create a static function in the interface. Something like this?

IPlatformLog.h
```
static std::unique_ptr<IPlatformLog> IPlatformLog::CreatePlatformInterface()
```

Then in each interface define the function?
PosixInterfaceForCLog.cpp
```
std::unique_ptr<IPlatformLog> CPosixInterfaceForCLog::CreatePlatformInterface()
{
  return std::make_unique<CPosixInterfaceForCLog>();
}
```

Montellese replied:
Either I don't understand what you mean or I don't see the benefit. At some point we will have to have `#ifdef`s to include the proper headers. I can only combine the includes with the instantiations to reduce the two `#ifdef` blocks to a single one. Having a static method in an interface doesn't make any sense.

lrusak replied:
Sure, it doesn't have to exist in the interface, but a base class on top would be fine. See CCPUInfo for an example of what I mean: https://github.com/xbmc/xbmc/blob/master/xbmc/utils/CPUInfo.h#L76
https://github.com/xbmc/xbmc/blob/master/xbmc/platform/linux/CPUInfoLinux.cpp#L65

Montellese replied:
Done, I've added `IPlatformLog::CreatePlatformLog()`.

lrusak commented:
I've been trying to get away from the three build methods. IMO we don't need to have the depends method that can be built directly. ie. (make -C tools/depends/target/spdlog`. Having the external project in cmake and the full depends method is enough IMO.

I'm also not sure if you want to try and do the auto fallback to the internal cmake build if we cannot find the library like how we do with ffmpeg and dav1d. That way we don't have to change the jenkins build instructions or the host libraries everytime a platform needs another lib.

Montellese commented:
What would that mean for `FindSpdlog.cmake` and `tools/depends/target/libspdlog/Makefile`? I can probably try to duplicate the behaviour for `dav1d` but do I need to make any other adjustments?

lrusak reviewed (commented):
> What would that mean for `FindSpdlog.cmake` and `tools/depends/target/libspdlog/Makefile`? I can probably try to duplicate the behaviour for `dav1d` but do I need to make any other adjustments?

For dav1d we just check to see if the libraries exist and if not fallback to the internal build, see: https://github.com/xbmc/xbmc/blob/master/cmake/modules/FindDav1d.cmake#L24-L27

and for the Makefile you don't have to change it but you could get rid of the PLATFORM check as that's only needed to detect if we are doing a full depends build or not.

Montellese commented:
> For dav1d we just check to see if the libraries exist and if not fallback to the internal build, see: https://github.com/xbmc/xbmc/blob/master/cmake/modules/FindDav1d.cmake#L24-L27

Already saw that and I'm giving it a try now. But a problem might be that the version shipped with Ubuntu 16.04 and 18.04 will be detected but will not be compatible. Should we
- enable internal spdlog build by default or
- introduce a version check and force and error or switch to internal build if the version is too old? The problem with this is that Ubuntu uses the version `1.6` which is newer than any real `spdlog` release...

fuzzard commented:
Heres just some reference info regarding the tools/depends change Montellese has just put in.

With just the header copy for macos the findspdlog.cmake gives
```
-- Found PCRE: /Users/Shared/xbmc-depends/macosx10.15_x86_64-target-debug/lib/libpcrecpp.a (found version "8.42") 
-- Found RapidJSON: /Users/Shared/xbmc-depends/macosx10.15_x86_64-target-debug/include (found version "1.1.0") 
-- Found Spdlog: /Users/Shared/xbmc-depends/macosx10.15_x86_64-target-debug/include  
-- Found Sqlite3: /Users/Shared/xbmc-depends/macosx10.15_x86_64-target-debug/lib/libsqlite3.a (found version "3.26.0") ```

The headers are copied as expected. Build succeeds with no problems.
Gave it a runtime test as well, and looks ok. I enabled component debuggin of ffmpeg and network, so you can see what it outputs. Ive trimmed it (cut the last 25k lines), as ffmpeg does some serious logging, haha

https://paste.kodi.tv/yiyanateyi.kodi

Memphiz has mentioned hes already changed the case for the log type (info -> INFO, debug -> DEBUG)

fuzzard commented:
Regarding the ubuntu situation, as its a header only use currently, id almost lean towards enabling internal by default for all platforms, but ill leave that to you guys to decide

Montellese commented:
I think I found the proper solution for the dependency problem of `python_ binding`. I have to add `spdlog` to `GLOBAL_TARGET_DEPS` and then everything is fine. But the question remains whether to enable internal spdlog (header only) build by default or not?

fuzzard commented:
did you want to use the compiled lib at all, or is the header only use how you want to proceed at the moment?

if its header only, id just use internal always. Can always evaluate whether use of the compiled lib becomes desirable down the track.

Montellese commented:
> did you want to use the compiled lib at all, or is the header only use how you want to proceed at the moment?

The compiled lib only improves compile time because it specifically instantiates most of the necessary templates in the compiled library so the compiler doesn't have to do it again when compiling Kodi. If we see a major increase in compile time we can think about building and using the compiled library but for now I would start with the header only approach.

fuzzard commented:
sounds reasonable to me

Montellese commented:
jenkins build this please

gabime commented:
I strongly recommend using the compiled version. It reduces the compilation times significantly.

Montellese committed:
[cmake] fix FindFmt.cmake for windows and fix version from 3.0.1 to 6.1.2

Montellese commented:
jenkins build this please

Montellese commented:
OK looks like I got most platforms covered with
* building the compiled version of `spdlog` for faster compilation (thanks @gabime)
* automatically detecting if `spdlog` is missing and issuing an internal build

Now I only need to build and re-pack `spdlog` for
* Win32
* UWP Win64
* UWP Win32
* UWP ARM

Who is the resident specialist on UWP? I tried to configure / install my Visual Studio 2019 for UWP but running the `cmake` commands from `README.windows.md` still doesn't work...

jjd-uk commented:
@Paxxi has done some of the recent depends stuff for UWP

Paxxi commented:
https://github.com/Paxxi/kodi-deps/tree/work?files=1 is the current stuff for depends, haven't gotten around to merging it back yet.

It includes some powershell scripts to build and package for all platforms. I hope the readme is understandable 😀

It should be a fairly easy packaging job so I can have a look at  it

Montellese commented:
> https://github.com/Paxxi/kodi-deps/tree/work?files=1 is the current stuff for depends, haven't gotten around to merging it back yet.

Thanks your VS config was enough to get me going. I'm uploading the dependencies now and will let jenkins decide whether I did everything right or not.

Montellese commented:
Let's hope that the uploaded windows dependency packages have been distributed to the mirrors.

jenkins build this please

Montellese commented:
Looks like I was too early. Manually tested now that I can download the dependencies from mirrors.kodi.tv.

jenkins build this please

Montellese commented:
@Paxxi maybe you can help me. I uploaded the win32 packages for all platforms to the mirrors and the WIN-64 build is ok but the WIN-32 and UWP-WIN-64 builds fail because they can't download / extract the package, see e.g. https://jenkins.kodi.tv/job/WIN-UWP-64/9639/console. But I checked and manually downloaded the packages (e.g. http://mirrors.kodi.tv/build-deps/win32/spdlog-1.5.0-win10-x64-v141-2020317.7z for UWP-WIN-64) works fine.

Do you have any idea what could be wrong?

Montellese commented:
So the buildsystem integration is working now with CMake automatically falling back to building spdlog internally if not found.

Now the only question remaining is what to do about the discrepancy between `LOGINFO` and `LOGNOTICE` which is not supported by `spdlog`. The same goes `LOGSEVERE` and `LOGERROR` / `LOGFATAL`.

Paxxi commented:
I say squish them together to the equivalent spdlog level. I don't find the distinction between them useful in our scenario. In a server situation with log monitoring then sure, the difference between them might be very relevant.

Montellese committed:
[utils] add CStaticLoggerBase helper class for static loggers

Montellese committed:
move component logging from CAdvancedSettings to CLog

Montellese commented:
I've made the following changes to the PR:
* `CLog` is not a static global anymore. It can be instantiated, initialized, uninitialized and destroyed. It's "global" instance resides in `CServiceBroker` and can be retrieved using `CServiceBroker::GetLogging()`.
* I've moved the component specific logging code from `CAdvancedSettings` to `CLog`. No idea why it was in `CAdvancedSettings` in the first place. If I find the time I'll try to provide some example on how component specific logging could be simplified / improved with `spdlog`
* I've introduced the `CStaticLoggerBase` helper class which initializes a `static Logger s_logger` member with a specified name. Classes only requiring a static logger (e.g. in settings where it doesn't make sense to create a new logger for every instance) can derive from `CStaticLoggerBase` and simply use `s_logger` for logging directly using `spdlog`.

Montellese committed:
[addons] CAddonSettings: make use of new logging functionality

Montellese committed:
[network] HTTP request handlers: make use of new logging functionality

Montellese committed:
[network] UPnP: make use of new logging functionality

Montellese committed:
[network] CUPnPServer: remove unused SortItems() method

Montellese commented:
I've migrated some more classes to show how to use `spdlog` loggers. IMO this is it for this PR. We can always create new PRs for migrating other classes.

Montellese commented:
jenkins build this please

Paxxi commented:
Looks good to me.

Maybe we can create an up for grabs issue to convert the missing places? Maybe group by area or whatever way we have to keep it manageable.

I'm thinking it's a decent task for anyone curious about contributing but don't know where to start

Rechi commented on tools/depends/target/libspdlog/Makefile:
```diff
@@ -0,0 +1,67 @@
+ifeq ($(CROSS_COMPILING), yes)
+  DEPS += ../../Makefile.include
+else
+  CXXFLAGS += -std=c++11
```
Why C++11?

Montellese replied:
This is what I found in `crossguid` and it also seems to be used by `rapidjson`. Would you propose `c++14`? I couldn't find any depends build using that so far.

Rechi replied:
Doesn't the library itself select the correct C++ version it needs?
If it doesn't, C++14 should be used as that's the one Kodi uses.

Montellese replied:
No you manually have to specify whether your project support C++11 or C++14 and the library adapts depending on that. I can change it to C++14 it just doesn't fit what we use in other depends.

Montellese replied:
I've changed this to `-std=c++14`.

Rechi commented on tools/depends/target/libspdlog/Makefile:
```diff
@@ -0,0 +1,67 @@
+  endif
+endif
+
+LIBDYLIB=$(PLATFORM)/build/libfmt.a
```
Are you sure?

Montellese replied:
Nice spotting. Copy & paste mistake

Rechi commented on tools/depends/target/libspdlog/Makefile:
```diff
@@ -0,0 +1,67 @@
+endif
+	rm -rf $(PLATFORM); mkdir -p $(PLATFORM)
+	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
+	cd $(PLATFORM); rm -rf build; mkdir -p build
+	cd $(PLATFORM)/build; $(CMAKE) $(CMAKE_OPTIONS) ..
```
```Makefile
	cd $(PLATFORM); $(CMAKE) -B build $(CMAKE_OPTIONS)
```

Montellese replied:
This is how it's done for lots of other dependencies. A quick `grep` reveals
* libandroidjni
* libcec
* libfmt (where I copied this from)
* mariadb
* p8-platform
* rapidjson
* taglib

Only `libjpeg-turbo` uses your approach.

wsnipex replied:
this is really a minor. As Montellese points out, we historically use it as is and both do the same thing, so it's a matter of taste.

Rechi commented on cmake/modules/FindFmt.cmake:
```diff
@@ -12,6 +12,19 @@
+                          PATHS ${PC_FMT_INCLUDEDIR})
+
+if (NOT ENABLE_INTERNAL_FMT AND NOT FMT_INCLUDE_DIR)
+  set(ENABLE_INTERNAL_FMT ON)
```
Why auto fall-back to the internal build?
It should error if it is not found so user can then decide to install the (distribution) library or explicitly enable the internal build.

Montellese replied:
Please read the discussion in this PR. I asked for guidance on how to handle this and @lrusak said I should do it this way.

Rechi replied:
Using the distribution libraries is always preferred, except for libraries where Kodi requires a special version.
This isn't the case for this library so there is no reason for the auto fallback.

Montellese replied:
For Ubuntu the situation with `libfmt-dev` is as follows:
* 16.04 doesn't provide it at all.
* 18.04 provides an older incompatible version.
* 19.04 provides an older incompatible version.
* 20.04 will provide a compatible version.

So except for Ubuntu 20.04 we must use the version provided by Kodi. But please discuss this with @lrusak since he suggested to do this auto-fallback. I don't have any opinion on depends builds since I usually work on Windows.

wsnipex replied:
I've already backported libfmt to all our supported ubuntu series.
There is no need for this fallback

Montellese replied:
I reverted the whole commit. Now this PR only contains a fix for the expected version of `libfmt` / `Fmt`.

Rechi commented on cmake/modules/FindSpdlog.cmake:
```diff
@@ -0,0 +1,109 @@
+                             PATHS ${PC_SPDLOG_INCLUDEDIR})
+
+if (NOT ENABLE_INTERNAL_SPDLOG AND NOT SPDLOG_INCLUDE_DIR)
+  set(ENABLE_INTERNAL_SPDLOG ON)
```
Why auto fall-back to the internal build?
It should error if it is not found so user can then decide to install the (distribution) library or explicitly enable the internal build.

Montellese replied:
Please read the discussion in this PR. I asked for guidance on how to handle this and @lrusak said I should do it this way. Furthermore for spdlog there is currently no way to install the library through Ubuntu packages because the provided packages are too old and incompatible.

Rechi replied:
I'm not telling you to remove the internal build, only to remove the auto-fallback.
So if you haven't installed spodlog, your distribution doesn't provide spdlog or ships and older version them configure. Then you can decide to install it or configrure with internal spdlog, like it's currently possible for flatbuffers, fmt, fstrcmp and rapidjson.

Montellese replied:
For Ubuntu the situation with `libspdlog-dev` is as follows:
* 16.04 provides a version which doesn't actually exist. Nobody knows where it comes from. This version is incompatible.
* 18.04 provides an older incompatible version.
* 19.04 provides an older incompatible version.
* 20.04 will provide a compatible version.

So the situation is similar to `libfmt-dev`. But please discuss this with @lrusak since he suggested to do this auto-fallback. I don't have any opinion on depends builds since I usually work on Windows.

wsnipex replied:
I agree with Rechi, auto-fallback is not needed, specially since it doesn't work for packaging. Users should be aware what happens and decide for themselves how they want to proceed.

For ubuntu: I'll have to backport libspdlog first

Montellese replied:
I removed the fallback to internal build.

Montellese commented:
> Maybe we can create an up for grabs issue to convert the missing places? Maybe group by area or whatever way we have to keep it manageable.
> 
> I'm thinking it's a decent task for anyone curious about contributing but don't know where to start

Sure sounds like a good idea.

howie-f commented:
> Maybe group by area or whatever way we have to keep it manageable.

i‘m having a few days off right now. you can hand over some area to me if you like

fuzzard commented:
jenkins build this please

Montellese commented:
Hm for some reason some build agents / jobs couldn't checkout 209a909f24edeb00f2235076e8cf891b87606c84 while others could.

fuzzard commented:
Yeah, think I merged something as it was starting.
Jenkins build this please

Montellese commented:
OK now it looks better except for some build issue on TVOS.

fuzzard commented:
Yeah, the tvos thing isn't related, must have been an autocomf download failure or something. 
One more Time, Jenkins build this please

fuzzard commented:
Haha, now iOS fails downloading autoconf. Does it count if we got every platform built over 2 Jenkins runs with no code changes... Haha

fuzzard commented:
One more shot at getting green on all at once.

Jenkins build this please

Montellese commented:
> One more shot at getting green on all at once.

Unfortunately not. Some IOS / TVOS build always fails either on automake or autoconf. No idea why this is so shaky.

fuzzard commented:
I'll leave it up to you if you want to merge as is. We know its built for all platforms successfully.

Personally ive had hit and miss mirror downloads recently a few times as well. Obviously doesnt help us resolve the issue, but oh well.

Rechi commented:
There is no need to think about merging it yet, there are still changes required.

lrusak reviewed (approved):
Merge it so we can all itest t during this alpha period! 😃 Great work @Montellese 👍

Montellese commented:
I'm getting mixed feedback here so not sure how to proceed. Let's see what jenkins means with the latest change.

Montellese commented:
Wow it really did happen: jenkins is green ;-)

wsnipex commented on tools/depends/target/libspdlog/Makefile:
```diff
@@ -0,0 +1,67 @@
+.PHONY: .installed-native
+
+all: .installed-$(PLATFORM)
+
```
```suggestion
download: $(TARBALLS_LOCATION)/$(ARCHIVE)
```

Montellese replied:
And what triggers the `download` target?

wsnipex replied:
this is mostly used for my packaging, only triggered manually and not used in the normal build process

Montellese replied:
Ah so this is in addition to the `$(TARBALLS_LOCATION)/$(ARCHIVE)` target.

Montellese replied:
Added

wsnipex reviewed (changes requested):
see comments inline.

wsnipex commented:
spdlog 1.5.0 is now available for ubuntu in out nightly ppa. Not yet tested though

Montellese commented:
@wsnipex for some reason I cannot respond to your inline comments directly so I'll do it here.

First off thanks for the review and feedback. And thanks for building libspdlog 1.5.0 for our PPA.

How do I get libfmt and libspdlog from our PPA. I have it installed and I also tried both `sudo apt build-dep kodi` and `sudo apt isntall libfmt-dev` but the former didn't install `libfmt-dev` at all and the latter installed the incompatible version from Ubuntu. In retrospect I'm not sure if I already had `libfmt-dev` installed from Ubuntu which may be why it wasn't installed from our PPA? That's why I always had to either run `sudo make -C tools/depends/target/libfmt PREFIX=/usr/local` or run `cmake` with `-DENABLE_INTERNAL_FMT=ON` to get it working. That's why I figured that automatically falling back to building it internally would make it easier for developers.

wsnipex commented:
please check that the PPA is added correctly. You should see the following versions:
```
$ apt policy libfmt-dev
libfmt-dev:
  Installed: 6.1.2+ds-2ubuntu1~bionic
  Candidate: 6.1.2+ds-2ubuntu1~bionic
  Version table:
 *** 6.1.2+ds-2ubuntu1~bionic 500
        500 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu bionic/main amd64 Packages
        100 /var/lib/dpkg/status
$ apt policy libspdlog-dev
libspdlog-dev:
  Installed: 1:1.5.0-1~bionic
  Candidate: 1:1.5.0-1~bionic
  Version table:
 *** 1:1.5.0-1~bionic 500
        500 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu bionic/main amd64 Packages
        100 /var/lib/dpkg/status
```

Once this is correct, simply install with apt install libfmt-dev libspdlog-dev

wsnipex commented:
I ran a very quick test built against spdlog from the PPA. Seems to work fine

Montellese commented:
> apt policy libspdlog-dev

OK it looks like removing and re-adding the PPA solves the problem. I've tested the following cases:
* `libfmt-dev` and `libspdlog-dev` installed from PPA --> build works fine
* `libfmt-dev` and `libspdlog-dev` not installed --> CMake fails to find `libfmt`
* `libfmt-dev` and `libspdlog-dev` not installed but internal builds enabled --> build works fine

Montellese commented:
OK I think I have addressed all concerns from @Rechi based on the feedback from @wsnipex now. Thanks for the support.

lrusak commented:
You'll need to update the jenkins build steps for freebsd now as it uses the system libs and doesn't have libspdlog
```
11:36:43 -- spdlog not found, falling back to internal build
11:36:43 -- Found Spdlog: /home/jenkins/workspace/FreeBSD/build/build/lib/libspdlog.a (found version "1.5.0") 
```

you'll have to add `-DENABLE_INTERNAL_SPDLOG=ON` here:
https://github.com/xbmc/xbmc/blob/master/tools/buildsteps/freebsd/configure-xbmc#L8

Montellese commented:
> You'll need to update the jenkins build steps for freebsd now as it uses the system libs and doesn't have libspdlog

How does this work for `libfmt` on FreeBSD?

Montellese commented:
> You'll need to update the jenkins build steps for freebsd now as it uses the system libs and doesn't have libspdlog
> 
> ```
> 11:36:43 -- spdlog not found, falling back to internal build
> 11:36:43 -- Found Spdlog: /home/jenkins/workspace/FreeBSD/build/build/lib/libspdlog.a (found version "1.5.0") 
> ```
> 
> you'll have to add `-DENABLE_INTERNAL_SPDLOG=ON` here:
> https://github.com/xbmc/xbmc/blob/master/tools/buildsteps/freebsd/configure-xbmc#L8

Done

lrusak commented:
> > You'll need to update the jenkins build steps for freebsd now as it uses the system libs and doesn't have libspdlog
> 
> How does this work for `libfmt` on FreeBSD?

```
freebsd:~ # pkg info | grep libfmt
libfmt-6.1.2                   Formatting library for C++
```

wsnipex commented on .travis.yml:
```diff
@@ -70,7 +70,7 @@ install:
       gettext git-core gperf libasound2-dev libass-dev libbluray-dev libbz2-dev libcap-dev libcdio-dev libcec4-dev libcrossguid-dev libcurl3
       libcurl4-openssl-dev libdbus-1-dev libegl1-mesa-dev libfmt3-dev libfontconfig-dev libfreetype6-dev libfribidi-dev libfstrcmp-dev libgif-dev libgl1-mesa-dev
```
```suggestion
      libcurl4-openssl-dev libdbus-1-dev libegl1-mesa-dev libfmt-dev libfontconfig-dev libfreetype6-dev libfribidi-dev libfstrcmp-dev libgif-dev libgl1-mesa-dev
```

wsnipex replied:
do we even still use travis? If yes, the nightly PPA would need to be added for libfmt and libspdlog dependecies

Montellese replied:
I have no idea but if we do it was probably already broken... Should I fix this anyway?

wsnipex replied:
Imho it's out of scope for this PR. Let's check if this is still used and if needed fix it in a separate PR

wsnipex commented on docs/README.Ubuntu.md:
````diff
@@ -116,7 +116,7 @@ If you get a `package not found` type of message with the below command, remove
 Install build dependencies manually:
 ```
-sudo apt install debhelper autoconf automake autopoint gettext autotools-dev cmake curl default-jre doxygen gawk gcc gdc gperf libasound2-dev libass-dev libavahi-client-dev libavahi-common-dev libbluetooth-dev libbluray-dev libbz2-dev libcdio-dev libp8-platform-dev libcrossguid-dev libcurl4-openssl-dev libcwiid-dev libdbus-1-dev libegl1-mesa-dev libenca-dev libflac-dev flatbuffers-dev libfontconf ...
+sudo apt install debhelper autoconf automake autopoint gettext autotools-dev cmake curl default-jre doxygen gawk gcc gdc gperf libasound2-dev libass-dev libavahi-client-dev libavahi-common-dev libbluetooth-dev libbluray-dev libbz2-dev libcdio-dev libp8-platform-dev libcrossguid-dev libcurl4-openssl-dev libcwiid-dev libdbus-1-dev libegl1-mesa-dev libenca-dev libflac-dev flatbuffers-dev libfontconf ...
````
```suggestion
sudo apt install debhelper autoconf automake autopoint gettext autotools-dev cmake curl default-jre doxygen gawk gcc gdc gperf libasound2-dev libass-dev libavahi-client-dev libavahi-common-dev libbluetooth-dev libbluray-dev libbz2-dev libcdio-dev libp8-platform-dev libcrossguid-dev libcurl4-openssl-dev libcwiid-dev libdbus-1-dev libegl1-mesa-dev libenca-dev libflac-dev flatbuffers-dev libfmt-dev libfontconfig-dev libfreetype6-dev libfribidi-dev libfstrcmp-dev libgcrypt-dev libgif-dev libgles2-mesa-dev libgl1-mesa-dev libglu1-mesa-dev libgnutls28-dev libgpg-error-dev libiso9660-dev libjpeg-dev liblcms2-dev libltdl-dev liblzo2-dev libmicrohttpd-dev libmysqlclient-dev libnfs-dev libogg-dev libpcre3-dev libplist-dev libpng-dev libpulse-dev libshairplay-dev libsmbclient-dev libspdlog-dev libsqlite3-dev libssl-dev libtag1-dev libtiff5-dev libtinyxml-dev libtool libudev-dev libva-dev libvdpau-dev libvorbis-dev libxmu-dev libxrandr-dev libxslt1-dev libxt-dev lsb-release nasm ninja-build python3-dev python3-pil python3-pip rapidjson-dev swig unzip uuid-dev yasm zip zlib1g-dev
```

Montellese replied:
I will add these (`libfmt-dev` and `meson`) in a separate commit since they are not directly related to this PR.

wsnipex reviewed (approved):
Some minors in docs/README as suggested, otherwise good to go

Montellese committed:
[docs] update list of manually installed packages on Ubuntu

Co-Authored-By: Wolfgang Schupp <w.schupp@a1.net>

Montellese commented:
OK I've addressed all comments from @wsnipex. For some reason there's no jenkins build queued?

Montellese commented:
So jenkins made it and everything looks fine except that on LINUX-64 an unrelated unit test `TestSysfsPath.SysfsPathTestLongString` failed. No clue why.

gabime commented:
The string creation (and memory allocation and copy) in https://github.com/xbmc/xbmc/blob/aa1a404a9d68466e43b53e8b745a622efda61148/xbmc/platform/darwin/utils/DarwinInterfaceForCLog.mm#L38

can be avoided using:

```c++
 m_formatter->format(msg, formatted);
formatted.push_back(‘\0’); // so be passed as c_str
NSLog(@"%s", formatted.data());
```

Montellese commented:
> The string creation (and memory allocation and copy) in
> 
> https://github.com/xbmc/xbmc/blob/aa1a404a9d68466e43b53e8b745a622efda61148/xbmc/platform/darwin/utils/DarwinInterfaceForCLog.mm#L38
> 
> can be avoided using:
> 
> ```c++
> m_formatter->format(msg, formatted);
> formatted.push_back(‘\0’); // so can be passed as c_str
> NSLog(@"%s", formatted.data());
> ```

Thanks for the input. TBH I had to play around quite a bit to get working code since I have no clue about Apple's Foundation API.

MilhouseVH commented:
I've tested this PR with spdlog-1.5.0 and libfmt-6.2.0, and I thought it might be useful to post some before/after logs (from Linux) for comparison.

Before (without spdlog): http://ix.io/2hpn
After (with spdlog): http://ix.io/2hpm

Apologies if this has been discussed already - so many comments!

* The date is missing from each log entry, this can be useful, at times - is the removal intentional?

* Is the user (`root` in this case) on most log entries relevant? The presence of the current user isn't consistent, so may be an artefact of some kind

* Multi-line items such as `advancedsettings.xml` are output very differently from before, which may be OK, I suppose. :)

Montellese commented:
> * The date is missing from each log entry, this can be useful, at times - is the removal intentional?

No, IIRC the date wasn't there in the Kodi logs when I initially implemented spdlog support. I will add it ASAP.

> * Is the user (`root` in this case) on most log entries relevant? The presence of the current user isn't consistent, so may be an artefact of some kind

This is not the user. spdlog supports logger names so I had to set a name for the default logger and I chose `root`. Suggestions are welcome. You can see other logger names like `CSettingsManager` or `CWebServer[8080]` and so on. The idea behind this is that it should make it much easier to identify where a log message is coming from without having to remember to make it part of the log message itself.

> * Multi-line items such as `advancedsettings.xml` are output very differently from before, which may be OK, I suppose. :)

I'd have to check the old code to see how this was done. Should I investigate?

Paxxi commented:
Maybe general would work instead of root?

Montellese commented:
> Maybe general would work instead of root?

Yeah or just `kodi`?

Concerning the identiation of multi-line logs I found https://github.com/xbmc/xbmc/pull/17498/commits/b6b89a30fdc302826834c461d5821886fc026558#diff-b7ab0e9f01171ac7da6f30e8337abebfL198. Should I re-integrate this?

Montellese commented:
I've added the date and changed the default logger name from `root` to `general`. Now it looks like this:
```
2020-04-10 19:17:54.513 T:6580  WARNING CSettingsManager: missing version attribute
2020-04-10 19:17:54.523 T:6580     INFO general: Default Video Player: VideoPlayer
2020-04-10 19:17:54.523 T:6580     INFO general: Default Audio Player: paplayer
2020-04-10 19:17:54.523 T:6580     INFO general: Disabled debug logging due to GUI setting. Level 1.
2020-04-10 19:17:54.525 T:6580     INFO general: CMediaSourceSettings: loading media sources from special://masterprofile/sources.xml
2020-04-10 19:17:54.560 T:6580    DEBUG general: CSkinSettings: no <skinsettings> tag found
```
Is that ok?

lrusak commented:
@Montellese seems you have some tests failing

https://jenkins.kodi.tv/job/LINUX-64/16724/testReport/
https://jenkins.kodi.tv/job/WIN-64/12916/testReport/

Montellese commented:
> @Montellese seems you have some tests failing
> 
> https://jenkins.kodi.tv/job/LINUX-64/16724/testReport/
> https://jenkins.kodi.tv/job/WIN-64/12916/testReport/

Ah two of those fail because I changed the name of the default logger. But I have no idea why ` TestSysfsPath.SysfsPathTestLongString` is failing. It doesn't seem to have anything to do with logging.

MilhouseVH commented:
@Montellese 

> I've added the date and changed the default logger name from root to general

Thanks, the date looks good.

Multiple loggers could be useful.

Just wondering aloud, but would it make any sense to enclose the logger name in some sort of delimiter, eg. chevrons ie. `<general>`
```
2020-04-10 19:17:54.513 T:6580  WARNING <CSettingsManager> missing version attribute
2020-04-10 19:17:54.523 T:6580     INFO <general> Default Video Player: VideoPlayer
2020-04-10 19:17:54.523 T:6580     INFO <general> Default Audio Player: paplayer
```
Assuming we wanted to grep for a particular logger, a delimited logger name might make the task a little easier? It might also make it easier to differentiate the logger name from the logged message.

> Concerning the identiation of multi-line logs I found b6b89a3#diff-b7ab0e9f01171ac7da6f30e8337abebfL198. Should I re-integrate this?

I suppose the "old" indented approach is what we're all used to, so if it's not too much trouble that would be my choice, but if others disagree I'm not going to argue!

Montellese commented:
@lrusak I've fixed the two failed tests but I have no idea what to do about the `TestSysfsPath` one.

@MilhouseVH 
I'm not sure if searching for `<CSettingsManager>` will yield any other results than searching for `CSettingsManager` because I don't expect anyone else using the term in log messages. But I changed it and introduced the indentation for multi-line logs. Now it looks like this:
```
2020-04-11 11:04:34.397 T:22036    INFO <general>: Loaded settings file from special://profile/advancedsettings.xml
2020-04-11 11:04:34.398 T:22036    INFO <general>: Contents of special://profile/advancedsettings.xml are...
                                                   <advancedsettings>
                                                     <loglevel hide="false">1</loglevel>
                                                     <videolibrary>
```

fuzzard commented:
Jenkins build this please

MilhouseVH commented:
@Montellese thanks, that looks great to me!

fuzzard commented:
Great that we got an all green build, and i think all of @Rechi's comments have been fixed?

I dont want to give you more work, but @Montellese could you possibly apply the diff patches  at the latest jenkins job https://jenkins.kodi.tv/job/BuildMulti-PR/15524/artifact/

They are all clang-format diffs. Probably easiest to rebase and apply each commit diff as you go.

Any last comments from anyone else?

Rechi commented:
Please follow the code guidelines, at least for new code.

Montellese commented:
I've taken a look at some of the diffs created by clang-format and in some cases I'm confused. The following code is wrong (apart from being too long)
```cpp
void AddSinks(std::shared_ptr<spdlog::sinks::dist_sink<std::mutex>> distributionSink) const override {}
```
and is fixed to
```cpp
void AddSinks(std::shared_ptr<spdlog::sinks::dist_sink<std::mutex>> distributionSink) const override
{
}
```
which looks totally ridiculous in a header file and makes it unnecessarily verbose and long.

But the following code (in a CPP file) is also wrong
```cpp
~Testlog() override
{
  CServiceBroker::GetLogging().Uninitialize();
}
```
and is fixed to
```cpp
~Testlog() override { CServiceBroker::GetLogging().Uninitialize(); }
```

Where is the logic in that?

I will apply the clang-format diffs even though from the 8 diffs only one (partially) applies to new code. The rest is all just code that has been partially adjusted. I will however refuse to take responsibility for any of these changes especially if they result in contradicting code styles as displayed above.

Montellese committed:
~ follow coding style for d7489e5714

Montellese committed:
~ follow coding style for 50ced18d7b

Montellese committed:
~ follow coding style for 5af7aaeecc

Montellese committed:
~ follow coding style for 5163bf37fb

Montellese committed:
~ follow coding style for e0f2a73316

Montellese committed:
~ follow coding style for 362f28668f

Montellese committed:
~ follow coding style for 7b5518da91

Montellese committed:
~ follow coding style for ec6cd6ee41

Montellese commented:
OK I have applied the clang-format diffs from jenkins as follows:
* I have squashed changes into new / completely rewritten files.
* I have added extra code formatting commits after the original commits for all other changes where I only touched part of existing code.

lrusak commented:
jenkins build this please

lrusak commented:
Jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

Montellese commented:
I'm not sure why the OSX build failed. Looks like it couldn't find the `dmg` artifact at the end of the build?

lrusak commented:
> I'm not sure why the OSX build failed. Looks like it couldn't find the `dmg` artifact at the end of the build?

unfortunately it seems to be a random failure that we haven't tracked down yet :frowning_face:

Montellese commented:
I've rebased the branch, again...

I'm getting tired of this not moving forward especially if it is due to random failures on jenkins which have nothing to do with this PR. I just wanted to improve the general code of Kodi a bit but apparently that's not really on the agenda.

Paxxi commented:
Merge it. Any outstanding issues can be fixed separately

Montellese commented:
It looks like there's a problem configuring `spdlog` internally (`-DENABLE_INTERNAL_SPDLOG=ON`) or through depends (`make -C tools/depends/target/libspdlog/`) on Ubuntu if another version of `libfmt` is installed on the system. @MilhouseVH has this problem on Ubuntu 19.10 with `libfmt3-dev` installed. I was able to reproduce this on Ubuntu 18.04 with `sudo apt install libfmt-dev=4.0.0+ds-2`. The error message is
```
CMake Error at CMakeLists.txt:118 (add_library):
  Target "spdlog" links to target "fmt::fmt" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
```
in `spdlog`'s `CMakeLists.txt`.

Does anyone know why this could happen. Removing the system version results in a successful build so it does find the Kodi-specific version but maybe it first find the system version? How do we handle this case? Is this the user's / developer's mistake for having `libfmt` in the wrong version on the system?
But what about the `-DENABLE_INTERNAL_FMT=ON -DENABLE_INTERNAL_SPDLOG=ON` case where we explicitely know that `libfmt` has been built internally? Should we "hard code" where `spdlog` should look for `libfmt`?

wsnipex commented:
from a quick glance, it spdlog finds the system fmt first:
snip from spdlogs CMakeLists:
`    if (NOT TARGET fmt::fmt)
        find_package(fmt REQUIRED)
    endif ()
`
Our FindFmt module creates a target "fmt", maybe changing this to fmt::fmt would already fix the issue.

Montellese commented:
I already tried that and it didn't work. Same for @MilhouseVH...

HiassofT commented:
I noticed that build time and binary size of a debug builds in Linux about doubled since this PR got merged.

kodi 0fcbab8899 takes about 30 minutes to build on my laptop, kodi-x11 is about 220MB and total build folder size is 3.6GB
```
real	30m16.497s
user	107m44.543s
sys	7m57.897s
hias@lenny:~/private/libreelec/xbmc-nospdlog/build$ du -s -h .
3.6G	.
hias@lenny:~/private/libreelec/xbmc-nospdlog/build$ ls -lh kodi-x11 
-rwxr-xr-x 1 hias hias 222M Apr 23 15:33 kodi-x11
```
with commit 77a9d7c764 it's almost an hour, 370MB kodi-x11 binary and 11GB build folder size
```
real	57m32.700s
user	206m28.323s
sys	13m10.194s
hias@lenny:~/private/libreelec/xbmc-spdlog/build$ ls -lh kodi-x11 
-rwxr-xr-x 1 hias hias 373M Apr 23 15:35 kodi-x11
hias@lenny:~/private/libreelec/xbmc-spdlog/build$ du -s -h .
11G	.
```
I've built on Debian Buster and used internal spdlog and fmt (and made sure I purged libfmt-dev before) using these cmake options
```cmake .. \
	-DCMAKE_INSTALL_PREFIX=/opt/kodi \
	-DX11_RENDER_SYSTEM=gl \
	-DENABLE_INTERNAL_FLATBUFFERS=ON \
	-DENABLE_INTERNAL_FMT=ON \
	-DENABLE_INTERNAL_SPDLOG=ON \
	-DCMAKE_BUILD_TYPE=Debug \
	"$@"
```
Could you look into this?

With split debug info (DEBUGFISSION=ON) as in this build it's a bit annoying, but when building debug info into the binary (DEBUGFISSION=OFF) it's getting really problematic to even get a gdb backtrace from a debug kodi on devices without tons of RAM (like eg RPi)

HiassofT commented:
Here are the results with `-DENABLE_DEBUGFISSION=OFF`: about 500MB binary prior to this PR, 1.1GB binary with this PR
```
real	30m11.443s
user	109m12.409s
sys	8m2.992s
hias@lenny:~/private/libreelec/xbmc-nospdlog/build-nofission$ ls -lh kodi-x11 
-rwxr-xr-x 1 hias hias 529M Apr 23 18:36 kodi-x11
hias@lenny:~/private/libreelec/xbmc-nospdlog/build-nofission$ du -s -h .
5.2G	.
```
```
real	56m7.933s
user	199m42.135s
sys	12m32.319s
hias@lenny:~/private/libreelec/xbmc-spdlog/build-nofission$ ls -lh kodi-x11 
-rwxr-xr-x 1 hias hias 1.1G Apr 23 19:39 kodi-x11
hias@lenny:~/private/libreelec/xbmc-spdlog/build-nofission$ du -s -h .
14G	.
```

Montellese commented:
@HiassofT thanks for the report. Do you know if this also happens for `Release` builds?

TBH I never looked at the size of the binary and I also didn't observe compile time because my dev machine is so slow I usually do something else while running a full Kodi build.

I'll try to look into it but it won't be until sometime next week. If anybody has any good ideas / hints in the mean time feel free to speak up :-)

HiassofT commented:
Release builds just finished, with spdlog the kodi-x11 binary is about 10% larger but both build time and build folder size increased by about 50%.

without spdlog:
```
real	39m49.229s
user	149m45.630s
sys	7m16.450s
hias@lenny:~/private/libreelec/xbmc-nospdlog/build-release$ ls -lh kodi-x11 
-rwxr-xr-x 1 hias hias 47M Apr 24 14:45 kodi-x11
hias@lenny:~/private/libreelec/xbmc-nospdlog/build-release$ du -s -h .
828M	.
```
With spdlog:
```
real	60m25.610s
user	229m54.041s
sys	8m26.060s
hias@lenny:~/private/libreelec/xbmc-spdlog/build-release$ ls -lh kodi-x11 
-rwxr-xr-x 1 hias hias 51M Apr 24 15:46 kodi-x11
hias@lenny:~/private/libreelec/xbmc-spdlog/build-release$ du -s -h .
1.3G	.
```
configured with these options:
```
cmake .. \
	-DCMAKE_INSTALL_PREFIX=/opt/kodi \
	-DX11_RENDER_SYSTEM=gl \
	-DENABLE_INTERNAL_FLATBUFFERS=ON \
	-DENABLE_INTERNAL_FMT=ON \
	-DENABLE_INTERNAL_SPDLOG=ON \
	"$@"
```
